docs/testing/02-cli-tests.mdx

```diff
@@ -0,0 +1,123 @@
+schema. The schema is a basic application that stores and retrieves users:
+
+<details>
+  <summary> users.kf </summary>
+
+```sql showLineNumbers
+database users_db;
+
+table users {
+  id uuid primary key,
+  name text not null unique,
+  address text not null unique
+}
+
+procedure create_user ($name text) public  {
+  // derive a deterministic uuid from the blockchain transaction ID
+  // https://www.postgresql.org/docs/16.1/uuid-ossp.html#UUID-OSSP-FUNCTIONS-SECT
+  $uuid := uuid_generate_v5('f541de32-5ede-4083-bdbc-b29c3f02be9e'::uuid, @txid);
+
+  insert into users (id, name, address)
+  values ($uuid, $name, @caller);
+}
+
+procedure get_users() public view returns table (name text, address text) {
+  return select name, address from users;
+}
+```
+</details>
+
+## Defining A Test
+
+To define a test, create a new JSON file called `tests.json`. This test assumes we have the current directory
+setup:
+
+```shell
+working-dir/
+    |- tests.json
+    |- users.kf
+```
+
+The below file gives our tests a top-level name of "my users.kf tests". It will read in
+the schema in the file `users.kf`. It defines an insert statement to seed data into the
+schema named `users_db`, which is the name we defined for our schema in `users.kf`:
+
+```json title="tests.json"
+{
+    "name": "my users.kf tests",
+    "schema_files": ["users.kf"],
+    "seed_statements": {
+        "users_db": ["INSERT INTO users (id, name, address) VALUES ('42f856df-b212-4bdc-a396-f8fb6eae6901'::uuid, 'satoshi', '0xAddress')"]
+    },
+    "test_cases": []
+}
+```
+
+### Defining Individual Test Cases
+
+Within the `test_cases` array, you can define individual test cases. Each test case can
+define the following fields:
+
+- `name`: The name of the test case, which will be used to identify the test case if it fails.
+- `database`: The name of the database schema that you are targeting, as defined at the top of the schema.
+- `target`: The name of the action/procedure in the schema that you were testing.
+- `args`: An array of arguments that will be passed to the procedure. It must match the expected parameters of the procedure. It can be empty if the procedure does not take any arguments.
+- `returns`: A 2-dimensional array of expected return values. This matches the expected return table of the action/procedure. This can be empty if the procedure does not return anything, or if an error is expected.
+- `error`: If the test case is expected to return an error, you can define the expected error message here. The tool will check that the returned error contains this string as a substring. If empty, the test case is expected to succeed.
+- `caller`: The wallet address (or any other string) that will be assigned to `@caller`. If empty, the caller will be set to the same wallet address as the database deployer.
+- `height`: The height of the block that the test case will be run at. If empty, the height will be set to 0.
```

Comment:
Can all of these optional values be left off the JSON object, or do they need to be declared?

Reply:
They can be left off. See my example, it does not have all of them.